fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,281 @@
+    abigen!(SimpleCoin, "../contracts/SimpleCoin.abi");
+    // once we deploy this contract, this is the address we expect the contract to be deployed to
+    const CONTRACT_ADDRESS: &str = "f410fnz5jdky3zzcj6pejqkomkggw72pcuvkpihz2rwa";
+    // the amount we want to send to the contract
+    const SEND_BALANCE_AMOUNT: u64 = 1000;
+
+    let mut upgrade_scheduler = UpgradeScheduler::new();
+    upgrade_scheduler
+        .add(
+            Upgrade::new("mychain".to_string(), 1, |state| {
+                println!(
+                    "[Upgrade at height {}] Deploy simple contract",
+                    state.block_height()
+                );
+
+                // create a message for deploying the contract
+                let mut mf = MessageFactory::new_secp256k1(my_secret_key(), 1, state.chain_id());
+                let message = match mf
+                    .fevm_create(
+                        Bytes::from(
+                            hex::decode(CONTRACT_HEX)
+                                .context("error parsing contract")
+                                .unwrap(),
+                        ),
+                        Bytes::default(),
+                        TokenAmount::default(),
+                        GAS_PARAMS.clone(),
+                    )
+                    .unwrap()
+                {
+                    ChainMessage::Signed(signed) => signed.into_message(),
+                    _ => panic!("unexpected message type"),
+                };
+
+                // execute the message
+                let (res, _) = state.execute_implicit(message).unwrap();
+                assert!(
+                    res.msg_receipt.exit_code.is_success(),
+                    "{:?}",
+                    res.failure_info
+                );
+
+                // parse the message receipt data and make sure the contract was deployed to the expected address
+                let res = fvm_ipld_encoding::from_slice::<eam::CreateReturn>(
+                    &res.msg_receipt.return_data,
+                )
+                .unwrap();
+                assert_eq!(
+                    res.delegated_address(),
+                    Address::from_str(CONTRACT_ADDRESS).unwrap()
+                );
+
+                // lets upgrade the app version as well
+                let app_version = 1;
+
+                Ok(app_version)
+            })
+            .unwrap(),
+        )
+        .unwrap();
+
+    upgrade_scheduler
+        .add(
+            Upgrade::new("mychain".to_string(), 2, |state| {
+                println!(
+                    "[Upgrade at height {}] Sends a balance",
+                    state.block_height()
+                );
+
+                // build the calldata for the send_coin function
+                let (client, _mock) = ethers::providers::Provider::mocked();
+                let simple_coin = SimpleCoin::new(EthAddress::from_id(101), client.into());
+                let call = simple_coin.send_coin(
```

Comment:
I think that after it's been deployed you can even use the `ContractCaller` to make implicit calls to the thing. Here's an [example](https://github.com/consensus-shipyard/ipc/blob/6282ccc3f214b2de795dae4838c8f9ae7ffeea61/fendermint/testing/contract-test/src/ipc/subnet.rs#L63-L65) of using it to join a subnet by a non-system actor.